src/util/battery/batterylinux.cpp

```diff
@@ -1,105 +1,63 @@
 #include "util/battery/batterylinux.h"
 
-// Contains 'signals' so we undefine Qt's 'signals'.
-#undef signals
-#include <upower.h>
-
 #include <QtDebug>
 
 BatteryLinux::BatteryLinux(QObject* pParent)
     : Battery(pParent) {
+    client = up_client_new();
 }
 
 BatteryLinux::~BatteryLinux() {
+    g_object_unref(client);
 }
 
 void BatteryLinux::read() {
-    m_iMinutesLeft = Battery::TIME_UNKNOWN;
-    m_dPercentage = 0.0;
-    m_chargingState = Battery::UNKNOWN;
-
-    // NOTE(rryan): It would be nice if we could create the client
-    // once. However, while testing this up_client_get_devices(client) returned
-    // an empty list when I tried to re-use the UpClient instance.
-    UpClient* client = up_client_new();
     VERIFY_OR_DEBUG_ASSERT(client) {
       return;
     }
 
-#if !UP_CHECK_VERSION(0, 9, 99)
-    // Re-enumerate in case a device is added.
-    up_client_enumerate_devices_sync(client, NULL, NULL);
-#endif
-
-#if UP_CHECK_VERSION(0, 99, 8)
-    GPtrArray* devices = up_client_get_devices2(client);
-    VERIFY_OR_DEBUG_ASSERT(devices) {
-      return;
-    }
-#else
-    // This deprecated function doesn't set the free function for
-    // the array elements so we need to do it!
-    // https://bugs.freedesktop.org/show_bug.cgi?id=106740
-    // https://gitlab.freedesktop.org/upower/upower/issues/14
-    GPtrArray* devices = up_client_get_devices(client);
-    VERIFY_OR_DEBUG_ASSERT(devices) {
-      return;
+    guint kind;
+    gboolean isPresent;
+    guint state;
+    gdouble percentage;
+    gint64 timeToEmpty;
+    gint64 timeToFull;
+    UpDevice *device = up_client_get_display_device(client);
+    if (device) {
+        g_object_get(G_OBJECT(device),
+                     "kind", &kind,
+                     "is-present", &isPresent,
+                     "state", &state,
+                     "percentage", &percentage,
+                     "time-to-empty", &timeToEmpty,
+                     "time-to-full", &timeToFull,
+                     NULL);
+    } else {
+        isPresent = false;
```

Comment:
`isPresent` is not needed. Use an *early exit* strategy like for `m_client` to simplify the code, i.e. consistently apply the same pattern for the whole function.

Reply:
I believe `isPresent` should be checked because of hot pluggable batteries, see here:
https://upower.freedesktop.org/docs/Device.html#Device:IsPresent
(I don't know what `state` is reported in case of a battery device with no battery in the bay)

Reply:
Regarding the _early exit_ strategy, I'll wait until https://github.com/mixxxdj/mixxx/pull/2288 is resolved, and then apply the same strategy here.